sklearn/utils/_set_output.py

```diff
@@ -1,64 +1,183 @@
+            Original input dataframe. This is used to extract the metadata that should
+            be passed to `X_output`.
 
-    if isinstance(data_to_wrap, pd.DataFrame):
-        if columns is not None:
-            data_to_wrap.columns = columns
-        return data_to_wrap
+        columns : callable, ndarray, or None
+            The column names or a callable that returns the column names. The
+            callable is useful if the column names require some computation. If `None`,
+            then no columns are passed to the container's constructor.
+
+        Returns
+        -------
+        wrapped_output : container_type
+            `X_output` wrapped into the container type.
+        """
+
+    def is_supported_container(self, X):
+        """Return True if X are the supported container.
+
+        Parameters
+        ----------
+        X : container or list of containers
+            Containers to be checked.
+
+        Returns
+        -------
+        is_supported_container : bool
+            True if all items in `X` are the supported container.
+        """
+
+    def update_columns(self, X, columns):
+        """Update columns in X.
+
+        Parameters
+        ----------
+        X : container
+            Container which columns is updated.
+
+        columns : ndarray of str
+            Columns to container columns with.
+
+        Returns
+        -------
+        updated-X : container
+            Container with new names.
+        """
+
+    def hstack(self, Xs):
+        """Stack containers horizontally (column-wise).
+
+        Parameters
+        ----------
+        Xs : list of containers
+            List of containers to stack.
+
+        Returns
+        -------
+        stacked_Xs : container
+            Stacked containers.
+        """
+
+
+class PandasAdapter:
+    container_lib = "pandas"
+
+    def create_container(self, X_output, X_original, columns):
+        pd = check_library_installed("pandas")
+        columns = get_columns(columns)
+        index = X_original.index if isinstance(X_original, pd.DataFrame) else None
+
+        if isinstance(X_output, pd.DataFrame):
+            if columns is not None:
+                X_output.columns = columns
+            return X_output
+
+        return pd.DataFrame(X_output, index=index, columns=columns, copy=False)
+
+    def is_supported_container(self, X):
+        pd = check_library_installed("pandas")
+        if not isinstance(X, list):
+            X = [X]
+        return all(isinstance(item, pd.DataFrame) for item in X)
+
+    def update_columns(self, X, columns):
+        X.columns = columns
+        return X
+
+    def hstack(self, Xs):
+        pd = check_library_installed("pandas")
+        return pd.concat(Xs, axis=1)
 
-    return pd.DataFrame(data_to_wrap, index=index, columns=columns, copy=False)
+
+class PolarsAdapter:
+    container_lib = "polars"
+
+    def create_container(self, X_output, X_original, columns):
```

Comment:
Would it make sense to hace `def create_container(self, X_output, columns, index=None)` as a constructor?

Reply:
For xarray, it needs more metadata. For example, their rows and columsn can be given different names.

The other way to do it is:

```python
def create_contaienr(self, X_output, columns, metadata):
	...

def get_metadata(self, X):
	...
```

where `get_metadata` gets the metadata from the original input and then it gets passed on to `create_container`.

Reply:
This might be a bit more explicit that we want to get the metadata. But I don't want to block the PR on this aspect. This is currently experimental and I assume that we will revise the design when adding more container support.